### DISCUSSION TITLE
Lets Talk About Compute Shader 👾

### DISCUSSION THREAD

**Discussion by sDos280** [raysan5/raylib#2664] [Ideas]
Hello my fellow raylibers👋, I opened this discussion to ask the community what it thinks about Compute Shader.
As you know, as programmers 👨‍💻 we always try to make our games faster, one way I could think about making my games(or simulations) run more quickly is by harnessing the powers of the GPU by using Compute Shader 🤖.

In the last weeks, I have been trying to learn about implementing Compute Shader in raylib but got stuck in most places mostly by the lack of information. so I thought of making a simple example for **beginners** about Compute Shader so more people will understand the benefits of using Compute Shaders.

I will like your opinion about the topic.

**Comment by wlxmhls** [comment]
Shader language is a pain for me. Lots of mathmatics.

**Reply by wlxmhls** [reply]
There should be some tutorial of raylib shader programming.  There seems some default glsl shader variables, sometimes I hav no idea how they correspond to the C code.

**Reply by sDos280** [reply]
yeah, that right

**Reply by sDos280** [reply]
in some extent, shaders are usually used only for doing math operations in parallel

**Comment by ElectroidDes** [comment]
Don't use raylib - use godot. There are fully working compute shaders. Why do you need this raylib?

**Reply by sDos280** [reply]
Well, unlike godot rayib is a graphics library and not a game engine (which for me is easier to work with). I originally started this conversation to make a compute shader example for beginners so they may too use the power and the beauty of compute shaders...

**Comment by ElectroidDes** [comment]
> Well, unlike godot rayib is a graphics library and not a game engine (which for me is easier to work with). I originally started this conversation to make a compute shader example for beginners so they may too use the power and the beauty of compute shaders...

You can use Godot for Compute Shaders only. Unlike raylib, Godot can use both OpenGl and Vulcan for compute shaders. :)

**Comment by raysan5** [comment]
@sDos280 A compute shaders example was added some time ago: https://github.com/raysan5/raylib/blob/master/examples/others/rlgl_compute_shader.c

**Reply by sDos280** [reply]
Cool, I will check it out!

**Reply by sDos280** [reply]
Never mind, I thought this was another example (not the GOL example). 

So yeah, I am aware of this example but (when I was a beginner) it was hard for me to understand it so when I understood compute shades much better i wanted to create a more simpler example to show the power of compute shaders...

**Comment by rmcvid** [comment]
Hey everyone,

I'm trying to work with OpenGL through raylib, but whenever I use an EBO it doesn't work at all. For example, I wrote a small test based on the tutorial from https://learnopengl.com/
I'm trying to draw a square made of two triangles indexed with an EBO, but nothing shows up.
I've included the code below.
Does anyone have an idea why the square isn’t rendering?

```
float vertices[] = {
    -0.5f, -0.5f, 0.0f, //bot left
     0.5f, -0.5f, 0.0f, // bot right
    -0.5f,  0.5f, 0.0f, // top left
    0.5f,  0.5f, 0.0f   // top right
};
unsigned int indices[] = {
    0, 1, 2,  // first triangle: bot left, bot right, top left
    1, 3, 2   // second triangle: bot right, top right, top left
};

int main(void)
{
    InitWindow(800, 600, "Shader carree");
    SetTargetFPS(60);

    const char *vs = "#version 330 core\n"
        "layout (location = 0) in vec3 aPos;\n"
        "void main()\n"
        "{\n"
        "   gl_Position = vec4(aPos.x, aPos.y, aPos.z, 1.0);\n"
        "}\0";
    const char *fs = "#version 330 core\n"
        "out vec4 FragColor;\n"
        "void main()\n"
        "{\n"
        "   FragColor = vec4(1.0f, 0.5f, 0.2f, 1.0f);\n"
        "}\n\0";

    Shader shader = LoadShaderFromMemory(vs, fs);
    int locColor = GetShaderLocation(shader, "ourColor");


    // gen -- Load bind except for the vao one; 
    unsigned int vao = rlLoadVertexArray();                                  // generate but don't assign 
    unsigned int vbo = rlLoadVertexBuffer(vertices, sizeof(vertices), false);
    unsigned int ebo = rlLoadVertexBufferElement(indices, sizeof(indices), false);
    
    // no need to enable other it is done in the load function
    rlEnableVertexArray(vao);
    // set
    rlSetVertexAttribute(0, 3, RL_FLOAT, false, 3 * sizeof(float), 0);
    rlEnableVertexAttribute(0);

    int countFrame = 0;
    while (!WindowShouldClose())
    {
        BeginDrawing();
        ClearBackground(BLACK); 
        rlEnableShader(shader.id);
        rlEnableVertexArray(vao);
        //rlDrawVertexArray(0, 3);                   // from the vao it works but not from the ebo draw 3 vertices
        rlDrawVertexArrayElements(0,6,0);     // draw nothing
        rlDisableVertexArray();
        
        EndDrawing();

        countFrame++;
    }
    UnloadShader(shader);
    rlUnloadVertexArray(vao);
    rlUnloadVertexBuffer(vbo);
    rlUnloadVertexBuffer(ebo);
    CloseWindow();
    return 0;
}

```
